### PR TITLE
F-M1-Auth-Backend: Backend Implementation 

### PR DESCRIPTION
**Pull request by BanerjeeAgniva**
Admin Flow:

https://github.com/user-attachments/assets/344f5759-d312-444a-9ffd-e64409bdc753

# Pull Request

## Summary

---

## What Changed

M1 - Authentication and Authorization implemented  
---

## Why This Change?

M1 - Authentication and Authorization is required 

---

## Testing
- [x] Tested locally
- [x] Edge cases considered
- [x] No breaking changes introduced

### Test Evidence

image

image

image
j

---

### Areas Affected

- [x] Backend
- [ ] Frontend
- [ ] Routing
- [ ] Infra
- [x] Database
- [ ] NA

### Modules Affected

- [x] M1 — Auth
- [ ] M2 — Pricing
- [ ] M3 — Grid
- [ ] M4 — Orders
- [ ] M5 — Dispatch
- [ ] M6 — Routing
- [ ] M7 — Hub
- [ ] M8 — Barcode
- [ ] M9 — Airline
- [ ] M10 — SLA
- [ ] M11 — Exceptions
- [ ] NA

---

## Risks / Concerns

- 

---

## Checklist
- [ ] Code follows project conventions (package layout, no cross-module internal imports)
- [x] Self-review completed
- [ ] Append-only audit trail preserved (no mutations to scan/manifest/role-change records)
- [ ] No sensitive data or credentials exposed
- [x] Documentation updated if behaviour changed
- [x] Ready for review

### CODE REVIEW

File: common/src/main/java/com/oneday/common/domain/BaseEntity.java [hunk 1]
```diff
@@ -2,27 +2,26 @@
```
Comment (BanerjeeAgniva) [inline common/src/main/java/com/oneday/common/domain/BaseEntity.java:2]: Here's why each change makes sense:

1. @UuidGenerator → @GeneratedValue(strategy = GenerationType.UUID)
@UuidGenerator is Hibernate 6-specific. @GeneratedValue(strategy = GenerationType.UUID) is standard JPA 3.1 (Jakarta Persistence), making the code more portable and less tied to Hibernate internals.

2. @PrePersist → @CreationTimestamp
The manual @PrePersist callback sets createdAt on insert, but @CreationTimestamp does the same thing and is handled directly by Hibernate — less boilerplate, and you can't accidentally set it to the wrong value or forget the null check.

3. Added @UpdateTimestamp + updatedAt field
The original has no updatedAt, which means you can't tell when a record was last modified. @UpdateTimestamp makes Hibernate automatically stamp the field on every UPDATE, which is standard practice for any audit-friendly entity — and BaseEntity is meant to be the shared audit base for all entities in the monolith.

In short: the new version is more idiomatic (standard JPA over Hibernate-proprietary), removes manual lifecycle code, and adds the missing updatedAt audit column that every entity in a logistics platform needs.

File: auth/src/main/java/com/oneday/auth/domain/Role.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    @Column(name = "is_builtin", nullable = false)
+    private boolean builtin;
+
+    @Column(nullable = false)
+    private boolean active = true;
+
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JoinTable(
+            name = "role_permissions",
+            joinColumns = @JoinColumn(name = "role_id"),
```
Comment (BanerjeeAgniva) [inline auth/src/main/java/com/oneday/auth/domain/Role.java:35]: # Role-Permission ER Diagram

```mermaid
erDiagram
    Role {
        UUID id
        string name
        string displayName
        boolean cityScoped
        boolean builtin
        boolean active
    }

    Permission {
        UUID id
        string action
    }

    RolePermissions {
        UUID role_id
        UUID permission_id
    }

    Role ||--o{ RolePermissions : "role_id"
    RolePermissions }o--|| Permission : "permission_id"

File: auth/src/main/java/com/oneday/auth/domain/ApiKey.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@Getter
+@Setter
+@Entity
+@Table(name = "api_keys")
+public class ApiKey extends BaseEntity {
+
+    @Column(name = "key_hash", nullable = false, unique = true)
+    private String keyHash;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (BanerjeeAgniva) [inline auth/src/main/java/com/oneday/auth/domain/ApiKey.java:19]: This is a solid design for B2B. A single B2B company often has multiple internal systems hittination lets them rotateor revoke one without disrupting others.

File: auth/src/main/java/com/oneday/auth/domain/User.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @Column(nullable = false, unique = true)
+    private String email;
+
+    @Column(name = "password_hash", nullable = false)
+    private String passwordHash;
+
+    @Column(nullable = false)
+    private String name;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (BanerjeeAgniva) [inline auth/src/main/java/com/oneday/auth/domain/User.java:23]: Priya (role_id = "r3..."), Rohan (role_id = "r3..."), and Kavya (role_id = "r3...") all reference the same SUPERVISOR role, so many users point to one role.

File: auth/src/main/java/com/oneday/auth/api/AuthController.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    @PostMapping("/register")
+    public ResponseEntity<LoginResponse> register(@Valid @RequestBody RegisterRequest request) {
+        return ResponseEntity.ok(authService.register(request));
+    }
+
+    @PostMapping("/api-keys")
+    public ResponseEntity<ApiKeyCreateResponse> createApiKey(
+            @AuthenticationPrincipal AuthUserDetails principal,
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/api/AuthController.java:49]: I assume this is a part of springboot and gives you ability to get the user? Can you explain how it works, a short explanation would be fine

File: auth/src/main/java/com/oneday/auth/repository/ApiKeyRepository.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+import org.springframework.data.repository.query.Param;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public interface ApiKeyRepository extends JpaRepository<ApiKey, UUID> {
+
+    @Query("SELECT k FROM ApiKey k JOIN FETCH k.user u JOIN FETCH u.role WHERE k.keyHash = :keyHash AND k.active = true")
+    Optional<ApiKey> findActiveByKeyHashWithUser(@Param("keyHash") String keyHash);
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/repository/ApiKeyRepository.java:15]: What is this query and function used for?

File: auth/src/main/java/com/oneday/auth/dto/request/RoleChangeRequest.java [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package com.oneday.auth.dto.request;
+
+import jakarta.validation.constraints.NotNull;
+
+import java.util.UUID;
+
+public record RoleChangeRequest(
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/dto/request/RoleChangeRequest.java:7]: Will this role change request not have any additional parameters? Could you give an example of how this works with just a UUID and reason.

File: auth/src/main/java/com/oneday/auth/dto/request/ChangePasswordRequest.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.oneday.auth.dto.request;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public record ChangePasswordRequest(
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/dto/request/ChangePasswordRequest.java:6]: How is this different from resetpasswordrequest in terms of use case? I know changepasswordrequest but where will then resetpasswordrequest come into play?

File: auth/src/main/java/com/oneday/auth/api/AuthController.java (hunk 5 above)
Comment (BanerjeeAgniva) [reply auth/src/main/java/com/oneday/auth/api/AuthController.java:49]: @AuthenticationPrincipal extracts the authenticated user's principal (your AuthUserDetails) directly from Spring Security's SecurityContext and injects it into the method — no manual token parsing needed.

File: auth/src/main/java/com/oneday/auth/dto/request/ChangePasswordRequest.java (hunk 8 above)
Comment (BanerjeeAgniva) [reply auth/src/main/java/com/oneday/auth/dto/request/ChangePasswordRequest.java:6]: Image

File: auth/src/main/java/com/oneday/auth/dto/request/RoleChangeRequest.java (hunk 7 above)
Comment (BanerjeeAgniva) [reply auth/src/main/java/com/oneday/auth/dto/request/RoleChangeRequest.java:7]: Example end-to-end:

Request
PUT /users/7c9e6679-.../role
Authorization: Bearer <admin-jwt>

{
  "newRoleId": "a1b2c3d4-5678-...",
  "reason": "Promoted to City Ops Manager"
}

What the service does:
1. Extracts target user ID from path → 7c9e6679-...
2. Extracts actor ID from JWT principal → admin's UUID
3. Looks up the role by newRoleId → finds CITY_OPS_MANAGER
4. Saves the user with the new role
5. Appends a RoleAuditLog row: action=GRANT, actorId=<admin>, targetUserId=7c9e6679-..., reason="Promoted to City Ops Manager"

The reason is optional — you can send null or omit it entirely and it just logs the change without a comment.

File: auth/src/main/java/com/oneday/auth/api/UserController.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+    private final UserService userService;
+    private final AuthService authService;
+
+    public UserController(UserService userService, AuthService authService) {
+        this.userService = userService;
+        this.authService = authService;
+    }
+
+    @PostMapping
+    @PreAuthorize("hasAnyRole('ADMIN', 'STATION_MANAGER')")
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/api/UserController.java:35]: I am not very aware of how this works, i think preauthorise is a part of springboot, we store roles attached to a user in our db, how will hasanyrole() realise this value?

File: auth/src/main/java/com/oneday/auth/api/UserController.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+import java.util.List;
+import java.util.UUID;
+
+@RestController
+@RequestMapping("/users")
+public class UserController {
+
+    private final UserService userService;
+    private final AuthService authService;
```
Comment (Sidarthapati) [inline auth/src/main/java/com/oneday/auth/api/UserController.java:27]: I am assuming both of these services come in later PR's. Assuming a service layer over exisiting repo/domain layers.

File: auth/src/main/java/com/oneday/auth/repository/ApiKeyRepository.java (hunk 6 above)
Comment (BanerjeeAgniva) [reply auth/src/main/java/com/oneday/auth/repository/ApiKeyRepository.java:15]: Image

A business partner (like a logistics company) uses an API key instead of logging in with a password every time. When they make a request, the system looks up that key and immediately knows who they are and what they're allowed to do — all in one step. This test checks that lookup works correctly for an active, valid key.

**Review by Satvik2810 — DISMISSED** [review]

File: auth/src/main/java/com/oneday/auth/api/UserController.java (hunk 10 above)
Comment (BanerjeeAgniva) [reply auth/src/main/java/com/oneday/auth/api/UserController.java:27]: yessir, now its there, check

**Review by Sidarthapati — APPROVED** [review]
